tests/db_functions/test_dates.py

```diff
@@ -0,0 +1,611 @@
+from __future__ import unicode_literals
+
+from datetime import datetime
+from unittest import skipIf
+
+from django.conf import settings
+from django.db import connection
+from django.db.models import DateField, DateTimeField
+from django.db.models.functions.date import (
+    Extract, ExtractDay, ExtractHour, ExtractMinute, ExtractMonth,
+    ExtractSecond, ExtractWeekDay, ExtractYear, Trunc, TruncDay, TruncHour,
+    TruncMinute, TruncMonth, TruncSecond, TruncYear,
+)
+from django.test import TestCase, override_settings
+from django.utils import timezone
+
+from .models import DTModel
+
+try:
+    import pytz
+except ImportError:
+    pytz = None
+
+
+def microsecond_support(value):
+    return value if connection.features.supports_microsecond_precision else value.replace(microsecond=0)
+
+
+def truncate_to(value, kind, tzinfo=None):
+    def truncate(value, kind):
+        if kind == 'second':
+            return value.replace(microsecond=0)
+        if kind == 'minute':
+            return value.replace(second=0, microsecond=0)
+        if kind == 'hour':
+            return value.replace(minute=0, second=0, microsecond=0)
```

Comment:
This function is only used in the tests to make it easier to mimic what the database will return. I'll go back over it and check that there's no glaring inconsistencies. From memory I covered timezone conversions correctly.

(Melbourne is +10 now, but +11 at the start of the year)

trunc_year('2016-04-12 00:00:00', 'Melbourne') -> '2016-01-01 00:00:00' -> '2016-01-01 00:00:00 +11:00'